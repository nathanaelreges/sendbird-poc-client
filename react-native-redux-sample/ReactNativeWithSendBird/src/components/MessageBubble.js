import React from 'react';
import { View, Text } from 'react-native';
import Video from 'react-native-video';
import { FileItem, TextItem } from './MessageItem';
import { ImageItem } from './ImageItem';

const _renderNickname = nickname => {
  return nickname ? <Text style={{ fontSize: 9, color: '#7048e8', paddingBottom: 4 }}>{nickname}</Text> : null;
};

const _isImage = props => {
  return props.message.type.match(/^image\/.+$/);
};

const _isVideo = props => {
  return props.message.type.match(/^video\/.+$/);
};

const MessageBubble = props => {
  let content = null;
  const message = props.message;
  if (!message) {
    return null;
  }
  if (message.isUserMessage()) {
    content = (
      <View
        style={{ maxWidth: 250, padding: 8, borderRadius: 8, backgroundColor: props.isUser ? '#5F3DC4' : '#e6e6e6' }}
      >
        {props.isUser || !props.isShow ? null : _renderNickname(props.nickname)}
        <View style={{}}>
          <TextItem isUser={props.isUser} message={message.message} />
        </View>
        <View style={{ flexDirection: 'row', justifyContent: 'flex-end', paddingLeft: 8 }}>
          <Text style={{ fontSize: 8, color: props.isUser ? '#E9EBEF' : '#878d99' }}>{props.time}</Text>
        </View>
      </View>
    );
  } else if (message.isFileMessage()) {
    if (_isImage(props)) {
      content = <ImageItem isUser={props.isUser} message={message.url.replace('http://', 'https://')} />;
    } else if (_isVideo(props)) {
      content = (
        <View style={styles.videoContainer}>
          <Video source={{ uri: message.url }} style={styles.video} paused={true} controls={true} />
        </View>
      );
    } else {
      content = <FileItem isUser={props.isUser} message={message.name} />;
    }
  }
  return content;
};

const styles = {
  videoContainer: {
    minWidth: 240,
    minHeight: 180,
    marginBottom: 4,
    borderRadius: 8
  },
  video: {
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    borderRadius: 8
  }
};

export { MessageBubble };
