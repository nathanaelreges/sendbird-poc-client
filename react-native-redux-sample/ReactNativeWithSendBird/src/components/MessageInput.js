import React from 'react';
import { View, Dimensions, Platform } from 'react-native';
import { Icon, FormInput } from 'react-native-elements';

const { width } = Dimensions.get('window');

const MessageInput = (props) => {
    /*
    <Icon
        containerStyle={{marginLeft: 14}}
        iconStyle={{margin: 0, padding: 0}}
        name='plus'
        type='font-awesome'
        color={'#494e57'}
        size={15}
        onPress={props.onLeftPress}
    />
    */
    return (
        <View style={{flexDirection: 'row'}}>
            <FormInput 
                containerStyle={{marginLeft: 8, marginRight: 8}}
                inputStyle={{
                    color: '#212529', 
                    ...Platform.select({
                        android: {
                            minHeight: 36,
                            width: width - 70,
                        },
                        ios: {
                            minHeight: 36,
                            width: width - 70,
                        },
                    })
                }}
                placeholder={'Your message'}
                autoCapitalize='none'
                autoCorrect={false}
                selectionColor={'#212529'}
                value={props.textMessage}
                onChangeText={props.onChangeText} 
                onFocus={props.onFocus}
            />
            <Icon
                containerStyle={{marginLeft: 0}}
                iconStyle={{margin: 0, padding: 0}}
                name='envelope'
                type='font-awesome'
                color={props.textMessage.length > 0 ? '#7d62d9' : '#494e57'}
                size={15}
                onPress={props.onRightPress}
            />
        </View>
    )
}

const styles = {

}

export { MessageInput }
