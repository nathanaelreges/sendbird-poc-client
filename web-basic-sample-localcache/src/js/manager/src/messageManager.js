
import MessageCollection from './collection/messageCollection';
import { MessageContainer } from './store/messageContainer';
import { MessageChunkContainer } from './store/chunkContainer';
import { MessageBroadcast } from './broadcast/messageBroadcast';

let _messageManager = null;
const _messageCollections = new WeakMap();
const _chunkContainer = new WeakMap();
const _messageContainer = new WeakMap();
const _broadcast = new WeakMap();

const CONNECTION_HANDLER_KEY = 'syncManager_messageManager_connectionHandler_' + new Date().getTime();
const CHANNEL_HANDLER_KEY = 'syncManager_messageManager_channelHandler_' + new Date().getTime();

export default class MessageManager {
  constructor(sendBird) {
    if(!_messageManager) {
      _messageManager = this;
      
      const sb = this.sb = sendBird;
      const messageContainer = new MessageContainer(sb);
      const chunkContainer = new MessageChunkContainer(sb);
      const broadcast = new MessageBroadcast();

      _messageCollections.set(this, []);
      _messageContainer.set(this, messageContainer);
      _chunkContainer.set(this, chunkContainer);
      _broadcast.set(this, broadcast);

      const connectionHandler = new sb.ConnectionHandler();
      connectionHandler.onReconnectStarted = () => {
        broadcast.pause();
      };
      connectionHandler.onReconnectSucceeded = () => {
        broadcast.reset();
      };
      sb.addConnectionHandler(CONNECTION_HANDLER_KEY, connectionHandler);

      const channelHandler = new sb.ChannelHandler();
      const eventKeys = Object.keys(channelHandler);
      eventKeys.forEach(event => {
        channelHandler[event] = (...args) => {
          const collections = _messageCollections.get(this);
          if(Array.isArray(collections)) {
            switch(event) {
              case 'onMessageReceived': {
                const message = messageContainer.upsert(args[1]);
                broadcast.upsert(message);
                break;
              }

              case 'onReadReceiptUpdated': {
                const channel = args[0];
                broadcast.read(channel);
                break;
              }

              case 'onMessageUpdated': {
                const message = messageContainer.upsert(args[1]);
                broadcast.update(message);
                break;
              }

              case 'onMessageDeleted': {
                const messageId = messageContainer.remove(parseInt(args[1]));
                broadcast.remove(messageId);
                break;
              }
            }
          }
        };
      });
      sb.addChannelHandler(CHANNEL_HANDLER_KEY, channelHandler);
    } else {
      return _messageManager;
    }
  }
  static get instance() {
    return _messageManager;
  }
  createMessageCollection(channel, filter) {
    const broadcast = _broadcast.get(this);
    const collection = new MessageCollection({
      manager: this,
      broadcast,
      messageContainer: _messageContainer.get(this),
      chunkContainer: _chunkContainer.get(this),
      channel,
      filter
    });
    if(collection) {
      broadcast.addCollection(collection);
    }
    return collection;
  }
  removeMessageCollection(collection) {
    const broadcast = _broadcast.get(this);
    broadcast.removeCollection(collection);
  }
  clearCache() {
    const messageContainer = _messageContainer.get(this);
    const chunkContainer = _chunkContainer.get(this);
    messageContainer.clear();
    chunkContainer.clear();
  }
  syncChangeLog(channel) {
    if(localStorage) {
      const messageContainer = _messageContainer.get(this);
      const broadcast = _broadcast.get(this);

      const tokenKey = `last-changeLog-token-${channel.url}`;
      const lastChangeLogToken = localStorage.getItem(tokenKey) || null;
      channel.getMessageChangeLogsByToken(lastChangeLogToken, (result, err) => {
        if(this.sb.getErrorFirstCallback()) {
          const temp = result;
          result = err;
          err = temp;
        }
        if(!err) {
          for(let i in result.updatedMessages) {
            const currentMessage = messageContainer.getItem(result.updatedMessages[i].messageId);
            const message = messageContainer.upsert(result.updatedMessages[i]);
            const isDirty = !currentMessage || !currentMessage.isEqual(message);
            if(isDirty) {
              broadcast.update(message);
            }
          }
          for(let i in result.deletedMessageIds) {
            messageContainer.remove(result.deletedMessageIds[i]);
            broadcast.remove(result.deletedMessageIds[i]);
          }
          localStorage.setItem(tokenKey, result.token);
          if(result.hasMore) {
            this.syncChangeLog(channel);
          }
        }
      });
    }
  }
  start() {
    const broadcast = _broadcast.get(this);
    broadcast.reset();
  }
  stop() {
    const broadcast = _broadcast.get(this);
    broadcast.pause();
  }
  reset() {
    if(_messageManager) {
      _messageManager.clearCache();
      _messageManager.sb.removeConnectionHandler(CONNECTION_HANDLER_KEY);
      _messageManager.sb.removeChannelHandler(CHANNEL_HANDLER_KEY);
      _messageManager = null;
    }
  }
}