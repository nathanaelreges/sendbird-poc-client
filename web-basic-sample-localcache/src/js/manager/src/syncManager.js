
import LocalDB from './store/engine/localdb.min.js';
import ChannelManager from './channelManager';
import MessageManager from './messageManager';
import ChangeLog from './util/changeLog';
import SyncManagerException from './util/exception';

export default class SyncManager {
  constructor() {
  }
  static init() {
    /** this manager uses LocalDB as a storage engine.
       *  LocalDB requires to call LocalDB.config() to use the storage.
       *  For more information about the engine,
       *  see https://github.com/smilefam/LocalDB.
      */
    LocalDB.config();

    SyncManager.Channel = ChannelManager;
    SyncManager.Message = MessageManager;
    SyncManager.ChangeLog = ChangeLog;
    SyncManager.Exception = SyncManagerException;
  }
}