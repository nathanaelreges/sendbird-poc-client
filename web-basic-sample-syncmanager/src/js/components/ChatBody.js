import styles from '../../scss/chat-body.scss';
import { createDivEl, getDataInElement, removeClass, findMessageIndex } from '../utils';
import { Message } from './Message';
import { SendBirdAction } from '../SendBirdAction';
import { MESSAGE_REQ_ID } from '../const';

import SendBirdSyncManager from 'sendbird-syncmanager';

class ChatBody {
  constructor(channel) {
    this.channel = channel;
    this.readReceiptManageList = [];
    this.scrollHeight = 0;
    this.collection = null;
    this.limit = 50;
    this.element = createDivEl({ className: styles['chat-body'] });
    this._initElement();
  }
  
  _initElement() {
    if(this.collection) {
      this.collection.remove();
    }
    this.collection = new SendBirdSyncManager.MessageCollection(this.channel, { limit: this.limit });
    this.collection.limit = this.limit;

    const collectionHandler = new SendBirdSyncManager.MessageCollection.CollectionHandler();
    collectionHandler.onMessageEvent = (action, message) => {
      const messageElements = this.element.querySelectorAll('.chat-message');
      const keepScrollToBottom = this.element.scrollTop >= this.element.scrollHeight - this.element.offsetHeight;
      switch(action) {
        case 'insert': {
          const index = findMessageIndex(message, this.collection.messages);
          if(index >= 0) {
            const messageItem = new Message({ channel: this.channel, message });
            if(index === this.collection.messages.length) {
              this.element.appendChild(messageItem.element);
            } else {
              this.element.insertBefore(messageItem.element, messageElements[index]);
            }
            if ((message.isUserMessage() || message.isFileMessage()) && SendBirdAction.getInstance().isCurrentUser(message.sender)) {
              this.readReceiptManage(message);
            }
          }
          break;
        }
        case 'update': {
          const messageItem = new Message({ channel: this.channel, message });
          const currentItem = this._getItem(message.messageId, false);
          const requestItem = message.reqId ? this._getItem(message.reqId, true) : null;
          if(currentItem || requestItem) {
            this.element.replaceChild(messageItem.element, requestItem ? requestItem : currentItem);
          }
          break;
        }
        case 'remove': {
          if(message.messageId) {
            this.removeMessage(message.messageId);
          } else {
            this.removeMessage(message.reqId, true);
          }
          break;
        }
        case 'clear': {
          while (this.element.firstChild) {
            this.element.removeChild(this.element.firstChild);
          }
          break;
        }
      }
      if(keepScrollToBottom) {
        this.scrollToBottom();
      }
    };
    this.collection.setCollectionHandler(collectionHandler);

    this.element.addEventListener('scroll', () => {
      if (this.element.scrollTop === 0) {
        this.updateCurrentScrollHeight();
        this.collection.fetch('prev', () => {
          this.element.scrollTop = this.element.scrollHeight - this.scrollHeight;
        });
      }
    });
  }

  loadPreviousMessages(callback) {
    this.collection.fetch('prev', () => {
      if(callback) callback();
    });
  }

  scrollToBottom() {
    this.element.scrollTop = this.element.scrollHeight - this.element.offsetHeight;
  }

  updateCurrentScrollHeight() {
    this.scrollHeight = this.element.scrollHeight;
  }

  repositionScroll(imageOffsetHeight) {
    this.element.scrollTop += imageOffsetHeight;
  }

  updateReadReceipt() {
    this.readReceiptManageList.forEach(message => {
      if (message.messageId.toString() !== '0') {
        const className = Message.getReadReceiptElementClassName();
        const messageItem = this._getItem(message.messageId, false);
        if (messageItem) {
          let readItem = null;
          try {
            readItem = messageItem.getElementsByClassName(className)[0];
          } catch (e) {
            readItem = null;
          }
          const latestCount = SendBirdAction.getInstance().getReadReceipt(this.channel, message);
          if (readItem && latestCount.toString() !== readItem.textContent.toString()) {
            readItem.innerHTML = latestCount;
            if (latestCount.toString() === '0') {
              removeClass(readItem, className);
            }
          }
        }
      }
    });
  }

  readReceiptManage(message) {
    for (let i = 0; i < this.readReceiptManageList.length; i++) {
      if (message.reqId) {
        if (this.readReceiptManageList[i].reqId === message.reqId) {
          this.readReceiptManageList.splice(i, 1);
          break;
        }
      } else {
        if (this.readReceiptManageList[i].messageId === message.messageId) {
          this.readReceiptManageList.splice(i, 1);
          break;
        }
      }
    }
    this.readReceiptManageList.push(message);
    this.updateReadReceipt();
  }

  _getItem(messageId, isRequestId = false) {
    const items = this.element.childNodes;
    for (let i = 0; i < items.length; i++) {
      const elementId = isRequestId ? getDataInElement(items[i], MESSAGE_REQ_ID) : items[i].id;
      if (elementId === messageId.toString()) {
        return items[i];
      }
    }
    return null;
  }

  removeMessage(messageId, isRequestId = false) {
    const removeElement = this._getItem(messageId, isRequestId);
    if (removeElement) {
      this.element.removeChild(removeElement);
    }
  }
}

export { ChatBody };
