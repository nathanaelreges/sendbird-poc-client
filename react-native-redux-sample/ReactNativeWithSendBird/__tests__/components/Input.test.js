import 'react-native';
import React from 'react';
import { Input } from '../../src/components';
import renderer from 'react-test-renderer';

describe('component/Input', () => {
  it('renders correctly', () => {
    const props = {
      label: 'test_label',
      value: 'test_value',
      maxLength: 123,
      onChangeText: () => {}
    };
    const tree = renderer.create(<Input {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
