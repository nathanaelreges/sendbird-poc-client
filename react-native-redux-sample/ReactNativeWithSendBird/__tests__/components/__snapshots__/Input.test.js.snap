// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/Input renders correctly 1`] = `
<RCTScrollView>
  <View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      borderColor="#d9d9d9"
      duration={100}
      highlightColor="#7d62d9"
      keyboardType="default"
      label="test_label"
      maxLength={123}
      onChangeText={[Function]}
      textBlurColor="#000"
      textFocusColor="#7d62d9"
      underlineColorAndroid="transparent"
      value="test_value"
    />
  </View>
</RCTScrollView>
`;
